### PR TITLE
NaturalSort performance improvements

### PR DESCRIPTION
**Pull request by RobinMcCorkell**
Nice ~~7%~~ _15%_ performance boost (thanks @LukasReschke for the idea!)

https://blackfire.io/profiles/compare/1b74a344-74b4-44c1-a0c6-abdbf8be3e2b/graph

graph
graph2
graph3

### CODE REVIEW

**Comment by LukasReschke** [comment]
What data did you use for testing? – Has that an impact on https://github.com/owncloud/core/issues/13434: "We waste more than 30 seconds on sorting and comparing the file names" ?

**Comment by Xenopathic** [comment]
@LukasReschke ehh, I just copied all the files from my /etc into a folder readable by oC, then used that. It comes out to 2442 files, with a variety of file names (I wanted to give the parser a real workout :smile:)

**Comment by LukasReschke** [comment]
Let me test this then on my setup later

**Comment by Xenopathic** [comment]
@LukasReschke how did your testing go?

**Comment by LukasReschke** [comment]
@Xenopathic Later !== "in 7 hours" ;-)

**Comment by LukasReschke** [comment]
@owncloud-bot Retest this please

File: lib/private/naturalsort.php [hunk 1]
```diff
@@ -75,14 +79,12 @@ public function compare($a, $b) {
...
-		$alen = count($aa);
-		$blen = count($bb);
 
-		for ($x = 0; $x < $alen && $x < $blen; $x++) {
+		for ($x = 0; isset($aa[$x]) && isset($bb[$x]); $x++) {
 			$aChunk = $aa[$x];
 			$bChunk = $bb[$x];
 			if ($aChunk !== $bChunk) {
-				if (is_numeric($aChunk) && is_numeric($bChunk)) {
+				if ($aChunk[0] >= '0' && $aChunk[0] <= '9' && $bChunk[0] >= '0' && $bChunk[0] <= '9') {
```
Comment (LukasReschke) [inline lib/private/naturalsort.php]: Can you please document what the `''` magic does here? – For example in a comment above it?

**Comment by LukasReschke** [comment]
On my instance with having 20k files in a single folder this leads to a speedup of 4 seconds when accessing the file list.

**Comment by Xenopathic** [comment]
Just pushed a second commit, related (not a NaturalSort improvement but speeds up `list.php` by another 5%).

graph

**Comment by LukasReschke** [comment]
@owncloud-bot Retest this please.

File: lib/private/files/fileinfo.php [hunk 2]
```diff
@@ -159,11 +159,10 @@ public function getPermissions() {
 	 * @return \OCP\Files\FileInfo::TYPE_FILE|\OCP\Files\FileInfo::TYPE_FOLDER
 	 */
 	public function getType() {
-		if (isset($this->data['type'])) {
-			return $this->data['type'];
-		} else {
-			return $this->getMimetype() === 'httpd/unix-directory' ? self::TYPE_FOLDER : self::TYPE_FILE;
+		if (!isset($this->data['type'])) {
+			$this->data['type'] = $this->getMimetype() === 'httpd/unix-directory' ? self::TYPE_FOLDER : self::TYPE_FILE;
```
Comment (PVince81) [inline lib/private/files/fileinfo.php]: Can you add brackets around `$this->getMimeType() === 'httpd/unix-directory'` for readability ?

If you'd forget them in JS the behavior would be `('httpd/unix-directory')? ... : ...` which is wrong.
Better be sure :smile:

**Comment by PVince81** [comment]
@Xenopathic please fix the comment above. Other than that the code looks good.

**Comment by PVince81** [comment]
Nice improvement, still works :+1:

**Comment by LukasReschke** [comment]
Agreed  – let's get this in. :+1:

**Pull request merged**